pinot-spi/src/main/java/org/apache/pinot/spi/environmentprovider/PinotEnvironmentProvider.java

```diff
@@ -0,0 +1,42 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.environmentprovider;
+
+import java.util.Map;
+import org.apache.pinot.spi.env.PinotConfiguration;
+
+/**
+ *  Environment Provider interface implemented by different cloud providers to customize
+ *  the base pinot configuration to add environment variables & instance specific configuration
+ */
+public interface PinotEnvironmentProvider {
+
+  String INSTANCE_FAILURE_DOMAIN = "pinot.environment.instance.failureDomain";
+
+  /**
+   * Customize base pinot configuration to add environment variables & instance specific configuration
+   * @return custom pinot configuration map
+   */
+  Map<String, Object> getEnvironment();
```

Comment:
Why return a map from the Env provider? Map is how Pinot chooses to store objects into zookeeper instance config. If we return a map from here, Pinot does not know how to parse the map.
Instead of this, I would declare a specific call like getFailureDomain(). Also, another call getFailureDomain(NodeName). 
In the first case, it returns a string with the failure domain name of the node in which the call is made. 
In the second case, it returns a string with the failure domain name of the specified node.

Reply:
We are relying on parsing the map in the HelixServerStarter. Also the reason for returning a map is if we decide to add more cloud environment specific variables (for now its only failure domain) then we can leverage the map at Server Starter layer to make ZK instance config write/updates. We are maintaining a Map<String, Map<String, String>> which gets updated with the above environment properties and Pinot recognizes it to update the mapFields. So returning a Map from EnvironmentProvider should not be an issue.

Reply:
Let us say, we have two cloud implementations, returning their map that we write to instance config today:

Azure returns: {"failureDomain": "foo", "Mem": "1024GB"}
And cloudX returns {"failureDomain" : "bar", "Memory" : 64GB"}
Assume that at some point later we want to add functionality where pinot behaves depending on memory in the system (say, it routes more queries to servers having more memory, or allocatets more segments, or whatever).

What would pinot code look like? (and each provider will claim that they are already providing that information)

My guess is it would be like:
`
mem = map.get("MEM"); if (mem == null) {mem = map.get("MEMORY");}
`
Bad

Reply:
Summary:
- We want to move away from each environment provider returning a custom map of environment properties which then gets populated to the instance configs
- Will have separate method for each environment property that we want pinot to makes sense out of to update the instance configs.
- This approach of having separate interface will close the possible backdoor of populating the instance configs with garbage/unknown properties and only restrict to properties that pinot can make sense of.
- We will return default values for these properties in case an environment/cloud provider does not have the logic to fetch these properties in their core environment provider implementation.